R/expand.R

```diff
@@ -1,78 +1,139 @@
+#' "{2001-01,2001-02-02}", "2008-XX-31", "..2002-02-03", "2001-01-03.."))
 #' expand(d)
 #' @export
-expand.messydt <- function(x) {
-
-  x <- remove_qualifiers(x)
-
+expand.messydt <- function(x, approx_range = 3) {
+  x <- stringr::str_remove_all(x, "[:space:]")
+  x <- stringr::str_remove_all(x, "\\{")
+  x <- stringr::str_remove_all(x, "\\}")
+  x <- stringr::str_remove_all(x, "\\%")
+  x <- stringr::str_remove_all(x, "\\?") # Dubious dates are not expanded...
+  x <- expand_approximate(x, approx_range)
   x <- expand_unspecified(x)
   x <- expand_sets(x)
   x <- expand_ranges(x)
-
   x
 }
 
-remove_qualifiers <- function(dates) {
-  dates <- stringr::str_remove_all(dates, "\\?")
-  dates <- stringr::str_remove_all(dates, "\\~")
-  dates <- stringr::str_remove_all(dates, "\\%")
-  dates <- stringr::str_remove_all(dates, "[:space:]")
-  dates <- stringr::str_remove_all(dates, "\\{")
-  dates <- stringr::str_remove_all(dates, "\\}")
+expand_approximate <- function(dates, approx_range) {
+
+  dates <- ifelse(stringr::str_detect(dates, "^[:digit:]{4}\\~$"),
+                  paste0(dates, "-01-01"), dates)
+  dates <- ifelse(stringr::str_detect(dates, "^[:digit:]{4}-[:digit:]{2}\\~$"),
+                  paste0(dates, "-01"), dates)
+  dates <- ifelse(stringr::str_detect(dates,
+  "^[:digit:]{4}\\~-[:digit:]{2}-[:digit:]{2}$|
+  |^\\~[:digit:]{4}\\-[:digit:]{2}-[:digit:]{2}$"),
+                  paste0(as.Date(gsub("\\~", "", dates)) -
+                           (365 * approx_range), "..",
+                         as.Date(gsub("\\~", "", dates)) +
+                           (365 * approx_range)), dates)
+  dates <- ifelse(stringr::str_detect(dates,
+                                      "^[:digit:]{4}\\-[:digit:]{2}\\~
+                                      -[:digit:]{2}$"),
+                  paste0(as.Date(gsub("\\~", "", dates)) - (31 * approx_range),
+                         "..",
+                         as.Date(gsub("\\~", "", dates)) + (31 * approx_range)),
+                  dates)
+  dates <- ifelse(stringr::str_detect(dates,
+                                      "^[:digit:]{4}-[:digit:]{2}-
+                                      [:digit:]{2}\\~$"),
+                  paste0(as.Date(gsub("\\~", "", dates)) - approx_range, "..",
+                         as.Date(gsub("\\~", "", dates)) + approx_range), dates)
+  dates <- ifelse(stringr::str_detect(dates, "^\\.\\."),
+                  paste0(as.Date(gsub("\\.\\.", "", dates)) -
+                           (365 * approx_range),
+                         "..", gsub("\\.\\.", "", dates)), dates)
+  dates <- ifelse(stringr::str_detect(dates, "\\.\\.$"),
+                  paste0(gsub("\\.\\.", "", dates), "..",
+                         as.Date(gsub("\\.\\.", "", dates)) +
+                           (365 * approx_range)),
+                  dates)
+  dates <- ifelse(stringr::str_detect(dates, "^[:digit:]{4}-XX-[:digit:]{2}$"),
+                  paste0(gsub("XX", "01", dates), "..",
```

Comment:
I have perhaps found a better solution to dealing with XX and sets here, I will upload my changes in a bit, let me know what you think!